### ISSUE TITLE
Problem with using websocket in a class on Windows platform

### ISSUE THREAD

**Issue by smartchris84** [websocket-client/websocket-client#469]
I've been using websocket with no problem on Mac and Ubuntu.

But when I was trying to use websocket in a class on Windows where on_open, on_message, on_close, on_error callbacks are effectively class functions, there are errors indicating callbacks are of invalid format.

But this code works perfectly on Mac and Ubuntu.

My Windows machine is Windows 10 and python is 3.6.4 (Anaconda).

Here's the code.

	import threading
	import websocket
	import ssl
	import sys
	from time import sleep

	class MyWebSocket():
		def connect(self, wsURL="wss://www.bitmex.com/realtime?subscribe=quote:XBTUSD"):
			print("Starting thread")
			ssl_defaults = ssl.get_default_verify_paths()
			sslopt_ca_certs = {"ca_certs": ssl_defaults.cafile}
			self.ws = websocket.WebSocketApp(wsURL,
	                                         on_message=self.__on_message,
	                                         on_close=self.__on_close,
	                                         on_open=self.__on_open,
	                                         on_error=self.__on_error,
	                                         )
			self.wst = threading.Thread(target=lambda: self.ws.run_forever(sslopt=sslopt_ca_certs))
			self.wst.daemon = True
			self.wst.start()

			conn_timeout = 5
			while (not self.ws.sock or not self.ws.sock.connected) and conn_timeout:
				sleep(1)
				conn_timeout -= 1
			if not conn_timeout:
				print("Could not connect to WS! Exiting.")
				sys.exit(1)

		def __on_message(self, ws, message):
			print(message)
		def __on_open(self, ws):
			print("opened")
		def __on_close(self, ws):
			print("closed")
		def __on_error(self, ws, error):
			print("error")

	websocket.enableTrace(True)
	mysocket = MyWebSocket()
	mysocket.connect()

	while True:
		sleep(1)

And here's the output. 
```
Starting thread
--- request header ---
GET /realtime?subscribe=quote:XBTUSD HTTP/1.1
Upgrade: websocket
Connection: Upgrade
Host: www.bitmex.com
Origin: http://www.bitmex.com
Sec-WebSocket-Key: E518/Zulj3+mTSfQh0aGiw==
Sec-WebSocket-Version: 13
--- response header ---
HTTP/1.1 101 Switching Protocols
Date: Wed, 22 Aug 2018 15:53:08 GMT
Connection: upgrade
Set-Cookie: AWSALB=qwD5YB1Ix9IoSM/Kcs/PGmTLDJ0303FUlCkbibY3ogdhfmyJVr9Gm8S+bToDngo0rXRULJlCjz0ItOmgdPmVvSlSVJNrvk26B2vZTBBUGW3hdhH5iws34bMh2S8L; Expires=Wed, 29 Aug 2018 15:53:08 GMT; Path=/
Upgrade: websocket
Sec-WebSocket-Accept: n9qPXu14/l2Z/dZjMdwfTpED2aw=
Sec-WebSocket-Version: 13
WebSocket-Server: uWebSockets
Strict-Transport-Security: max-age=31536000; includeSubDomains

error from callback <bound method MyWebSocket.__on_open of <__main__.MyWebSocket object at 0x000002376143B6D8>>: __on_open() missing 1 required positional argument: 'ws'
  File "C:\Users\Chris\Anaconda3\envs\kmeans\lib\site-packages\websocket\_app.py", line 333, in _callback
    callback(*args)
error from callback <bound method MyWebSocket.__on_message of <__main__.MyWebSocket object at 0x000002376143B6D8>>: __on_message() missing 1 required positional argument: 'message'
  File "C:\Users\Chris\Anaconda3\envs\kmeans\lib\site-packages\websocket\_app.py", line 333, in _callback
    callback(*args)
error from callback <bound method MyWebSocket.__on_error of <__main__.MyWebSocket object at 0x000002376143B6D8>>: __on_error() missing 1 required positional argument: 'error'
  File "C:\Users\Chris\Anaconda3\envs\kmeans\lib\site-packages\websocket\_app.py", line 333, in _callback
    callback(*args)
send: b'\x88\x82Q\xb8G\x1aRP'
error from callback <bound method MyWebSocket.__on_close of <__main__.MyWebSocket object at 0x000002376143B6D8>>: __on_close() missing 1 required positional argument: 'ws'
  File "C:\Users\Chris\Anaconda3\envs\kmeans\lib\site-packages\websocket\_app.py", line 333, in _callback
    callback(*args)
Couldn't connect to WS! Exiting.

```

Is there anyone who knows the solution? I googled but found no solution at all.

I can fix the error by using callbacks that are not member functions of a class but I need to use the class member function as the callbacks.

Please help me out!

Thanks.

**Comment by MRSharff** [comment]
This is area is a little bit hairy at the moment, PR https://github.com/websocket-client/websocket-client/pull/468 broke this functionality very recently. It should be reverted shortly. Refer to https://github.com/websocket-client/websocket-client/issues/471


You'll notice the traceback mentions `line 333, in _callback`

Here is the line, with a few leading and trails lines for context

```
    def _callback(self, callback, *args):
        if callback:
            try:
                if inspect.ismethod(callback):
                    callback(*args)
                else:
                    callback(self, *args)
```
Let's take one of your callbacks, `__on_message(self, ws, message)`, for example. 
`if inspect.ismethod(callback)` checks whether the callback is a method. In Python, class functions are typically referred to as methods, while module functions are referred to as functions. Since your `__on_message` is a method, it will enter the first block and call `callback(*args)`. This check is done because `*args` will contain `self` of the class the method belongs to, `MyWebSocket` in your case, and we don't want the `self` referred to in `def _callback(self, callback, *args)` to be passed in.

TLDR:
Your __on_message (and others) should only take self and message as arguments since _callback will not pass the WebSocketApp in.

```

	def __on_message(self, message):
		print(message)
	def __on_open(self):
		print("opened")
	def __on_close(self):
		print("closed")
	def __on_error(self, error):
		print("error")
```

As to why it's working on Mac and Ubuntu, I can't be certain but I would check the version of websocket-client you have installed on them. (This is probably due to the changes being made around this area recently.)

**Comment by smartchris84** [comment]
Thanks, with 0.48.0, the websocket is working perfect.

**Comment by MRSharff** [comment]
No problem. Looks like PR #468 was reverted so it should be fine to update to latest if you want. v0.53.0

**Comment by salihkilic** [comment]
> TLDR:
> Your __on_message (and others) should not take ws as an argument since _callback will not pass the WebSocketApp in.
> 
> ```
> 
> 	def __on_message(self, message):
> 		print(message)
> 	def __on_open(self):
> 		print("opened")
> 	def __on_close(self):
> 		print("closed")
> 	def __on_error(self, error):
> 		print("error")
> ```

The example in the Readme does use it. Is the example wrong then?
```
def on_message(ws, message):
    print(message)

def on_error(ws, error):
    print(error)

def on_close(ws):
    print("### closed ###")
```

**Comment by MRSharff** [comment]
Those functions in the example are not instance methods, so they will be passed the Websocket object.
